### DISCUSSION TITLE
Are there official samples for testing API routes with Jest

### DISCUSSION THREAD

**Discussion by thesobercoder** [vercel/next.js#15166] [Help] [answered]
I was able to follow the [with-jest](https://github.com/vercel/next.js/tree/canary/examples/with-jest) example and set it up correctly but I am unable find any official sample to test the API routes. Any help would be highly appreciated.

**Comment by juanbzz** [comment, marked as answer]
I don't think there's an official example, but this is how I test a custom middleware I'm working on using [supertest](https://www.npmjs.com/package/supertest):

```js
import request from "supertest";
import http, { IncomingMessage, ServerResponse } from "http";
import { apiResolver } from "next/dist/next-server/server/api-utils";
import { NextApiRequest, NextApiResponse } from "next";
import { nextCsrf } from "../index";

describe("test api route", () => {
  // mock for `apiResolver`'s 5th parameter to please TS
  const apiPreviewPropsMock = {
    previewModeId: "id",
    previewModeEncryptionKey: "key",
    previewModeSigningKey: "key",
  };

  const requestListener = (req: IncomingMessage, res: ServerResponse) => {
    apiResolver(
      req,
      res,
      undefined,
      csrf((req: NextApiRequest, res: NextApiResponse) => {
        return res.status(200).json({ message: "Hello, world." });
      }),
      apiPreviewPropsMock
    );
  };

  const { csrf } = nextCsrf({});
  it("should setup a CSRF token", async () => {
    // If we receive a request without secret in a cookie we assume it's the first request to an API route

    const server = http.createServer(requestListener);
    const agent = await request.agent(server).post("/");

    expect(agent.header["set-cookie"][0]).toEqual(
      expect.stringMatching(/XSRF-TOKEN=(.+); Path=\/; HttpOnly/g)
    );

    expect(agent.text).toBe(JSON.stringify({ message: "Hello, world." }));
  });
});
```

[Source](https://github.com/j0lv3r4/next-csrf/blob/master/src/middleware/csrf.test.ts)

**Reply by thesobercoder** [reply]
@j0lv3r4 Thanks, I will try this out. Some officials docs would really help newcomers like me, without having to open a discussion or hunt for a million links.

**Reply by thesobercoder** [reply]
@j0lv3r4 Thanks a lot for the example code. I was able to get it up and running within my project. I had a follow-up question though, if I were connecting to MongoDB in one of the APIs using mongoose, what will be the best way to mock that.

**Reply by lfades** [reply]
@j0lv3r4 @thesobercoder Please keep in mind that internal tools used by Next.js will not be documented and may be changed at any point without notice.

To test Next.js applications it's better to do integration tests by running your Next.js app and then make a request to the endpoint you want to test. Tests like this one are abstractions that don't reflect how your app will really work in production, may break at any moment, and will not be able to detect real problems, only those coming from the abstraction.

A good combination is to do unit tests for functions/methods used by your API endpoints (but not to the endpoint itself), but integration tests are way more useful.

**Reply by lfades** [reply]
@thesobercoder As for the other question, create a new local mongo db to run the tests and connect your DB to the mongo uri returned by the local db when you start the app, this is possible if you make the mongo uri an environment variable (which is most likely the case already).

**Reply by juanbzz** [reply]
@lfades I'm aware of the risks of using the internal tools but since I'm testing HTTP requests, headers, responses, and cookies, I don't think the way an API route handles those will change.

And you are right, running a Next.js app instance is a better option, but then I have to roll abstractions like https://github.com/vercel/next.js/blob/canary/test/lib/next-test-utils.js to use on all my projects because otherwise, I'd use internal tools again.

**Reply by timneutkens** [reply]
> I'd use internal tools again.

It's not the same though, next-test-utils just boots up `next dev`, `next build` or `next start` as a command. This is the public API as the CLI won't change between versions.

**Reply by juanbzz** [reply]
@timneutkens oh, awesome! Then it'd make sense to create tests using `next-test-utils.js` to run a Next.js app instance?

**Reply by timneutkens** [reply]
When you copy the code yeah, should be fine!

**Reply by olaven** [reply]
Hi! Are there any full examples on how to do this? 😄 I.e. starting the full server and making requests to it.

**Reply by eric-burel** [reply]
Hello from 2021, it would be great to see an example in the new testing doc (https://nextjs.org/docs/testing). I find it always tedious to run as server programmatically before running a test, because the process spawning API of Node is not very intuitive + you need to check if the server is not already running, check for errors etc. etc., any official recommendation or helper would be more than welcome.

**Reply by eric-burel** [reply]
Related discussion for lost googlers: https://github.com/vercel/next.js/discussions/28173

**Reply by momegas** [reply]
Change to `import { apiResolver } from "next/dist/server/api-utils";` for Next 12

**Reply by hjr3** [reply]
I ran into an issue with the my test hanging indefinitely. It turns out that my [Custom Config](https://nextjs.org/docs/api-routes/api-middlewares#custom-config) to set the `bodyParser` to `false` was not being included . I had to do something like `Object.assign(csrf, { config });` so the apiResolver would honor my config.

**Reply by im-ghost** [reply]
What if the route is protected

**Reply by gitmahin** [reply]
Update import for Next v16
```
import { apiResolver } from 'next/dist/server/api-utils/node/api-resolver';
```

**Comment by kmjennison** [comment]
I'm also interested in an easier way to test API routes and would love some clearer documentation/guides from the Next team. Here's a project attempting to solve this problem:
https://github.com/Xunnamius/next-test-api-route-handler

**Comment by lirbank** [comment]
We've been using this method for a while but with Next.js 12.1.0 `apiResolver` is not exported anymore, so all API tests fail. Anyone heard if there are any official testing recommendations for Next.js API handlers?

**Reply by hjr3** [reply]
I ran into this as well. You now want: `import { apiResolver } from 'next/dist/server/api-utils/node';`

From here: https://github.com/vercel/next.js/blob/v12.1.0/packages/next/server/api-utils/node.ts

**Reply by lirbank** [reply]
Oh crap! I just update our test suite to use https://www.npmjs.com/package/node-mocks-http instead of supertest. Oh, well, thanks for the help in any case! 👍

**Comment by ryanhaticus** [comment]
**!! ATTENTION WHEN USING A CUSTOM BODY PARSER IN THE TEST ENVIRONMENT !!**

As @hjr3 reported, you have to lie to `apiResolver` in testing because the handler's default export doesn't bring along the `export const config`. Special thanks to @falsestatement

This was a huge time waster for us, but here's our COMPLETE implementation. I hope this saves you time!

```typescript
import { createServer, IncomingMessage, ServerResponse } from 'http';
import { NextApiHandler } from 'next';
import { apiResolver } from 'next/dist/server/api-utils/node';

export const getMockServer = (
  handler: NextApiHandler,
  query: Record<string, string> = {},
  bodyParser = true,
) => {
  const requestHandler = (
    request: IncomingMessage,
    response: ServerResponse<IncomingMessage>,
  ) =>
    apiResolver(
      request,
      response,
      query,
      Object.assign(handler, {
        config: {
          api: {
            bodyParser,
          },
        },
      }),
      {
        previewModeEncryptionKey: '',
        previewModeId: '',
        previewModeSigningKey: '',
      },
      true,
    );

  return createServer(requestHandler);
};
```

**Reply by blakesoley** [reply]
Oh my god, thank you so much!  This was ridiculously hard to find an answer to, and I lost a lot of time trying to figure out why our jest tests were hanging.

More specifically to feed the search engines here - we have a React app and we implemented file uploading using formidable, and tests using jest and supertest.  Our tests were hanging with jest timeouts - formidable would get as far as the first `form.on('progress', (bytesReceived, bytesExpected)` event where we would get 0 bytesReceived out of the total expected, and then nothing as it timed out.

This is the fix!  We had `bodyParser: false` in the Next.js route handler already of course, but we needed to add it to the apiResolver for superTest as @ryanhaticus did here.

**Reply by akshg05** [reply]
Hey, how do we test a particular api path?
We have an api in a project as /api/newItem.
I want to write a test where an incorrect api path should fail like /api/newItems and only /api/newItem should succeed.
Is there a way to achieve this?

**Comment by ymc9** [comment]
Hey, has anyone figured out how to do this for the new "app dir" style API routes 😄

**Reply by helloworld** [reply]
This worked for me:

`src/test-client.ts`:
```
import { createServer, IncomingMessage, ServerResponse } from "http";
import { NextApiHandler } from "next";
import { apiResolver } from "next/dist/server/api-utils/node";
import request from "supertest";
import { parse } from "url";

export const testClient = (handler: NextApiHandler, bodyParser = true) => {
  return request.agent(
    createServer((req: IncomingMessage, res: ServerResponse) => {
      const { query } = parse(req.url || "", true); // Default to an empty string if `req.url` is undefined

      const customConfig = {
        api: {
          bodyParser,
        },
      };

      return apiResolver(
        req,
        res,
        query,
        Object.assign(handler, { config: customConfig }),
        {
          previewModeEncryptionKey: "",
          previewModeId: "",
          previewModeSigningKey: "",
        },
        true
      );
    })
  );
};
```

`jest.config.js`:
```
const nextJest = require("next/jest");

// Providing the path to your Next.js app which will enable loading next.config.js and .env files
const createJestConfig = nextJest({ dir: "./" });

// Any custom config you want to pass to Jest
const customJestConfig = {};

// createJestConfig is exported in this way to ensure that next/jest can load the Next.js configuration, which is async
module.exports = createJestConfig(customJestConfig);
```

`src/user.test.ts`: 
```
import { GET } from "./app/api/user/route";
import { testClient } from "./test-client";

const request = testClient(GET);

describe("/user", () => {
  it("should return current user", async () => {
    const res = await request.get("/"); 
    expect(res.status).toBe(200);
    expect(res.body.name).toBeDefined();
  });
});
```

`/api/users/route.js`:
```
// To handle a GET request to /api
export async function GET(req, res) {
  res.status(200).json({ name: "Hello world" });
}
```
